text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,195 @@
+  - When people actually need this information (e.g. to check if they're impacted by a vulnerability) it is too late to reach for third-party tooling - the executables have already been built and deployed, and the information is already lost. As such, this mechanism is ineffective if it's not enabled by default.
+
+# Prior art
+[prior-art]: #prior-art
+
+An out-of-tree implementation of this RFC exists, see [`cargo auditable`](https://github.com/rust-secure-code/cargo-auditable/), and has garnered considerable interest. NixOS and Void Linux build all their Rust packages with it today; it is also used in production at Microsoft. Extracting the embedded data is already supported by [`rust-audit-info`](https://crates.io/crates/rust-audit-info) and [Syft](https://github.com/anchore/syft). Auditing such binaries for known vulnerabilities is already supported by [`cargo audit`](https://crates.io/crates/cargo-audit) and [Trivy](https://github.com/aquasecurity/trivy).
+
+The Rust compiler already [embeds](https://github.com/rust-lang/rust/pull/97550) compiler and LLVM version in the executables built with it.
+
+Go compiler embeds `go.mod` dependency information into its compiled binaries. Due to Go binaries generally being far larger than Rust binaries, the binary size is not a constraint, so they embed much more information - e.g. the licence for each package in the dependency tree, which is then read by the [golicense](https://github.com/mitchellh/golicense) tool.
+
+The most common way to manage Ruby apps involves `Gemfile.lock` which can be thought of as a runtime `Cargo.lock`. Some companies have automation searching for these files in production VMs/containers and cross-referencing them against [RubySec](https://rubysec.com/).
+
+Since build system and package management system are usually decoupled, most other languages did not have the opportunity to implement anything like this.
+
+In microservice environments it is fairly typical to expose an HTTP endpoint returning the application version, see e.g. [example from Go cookbook](https://blog.kowalczyk.info/article/vEja/embedding-build-number-in-go-executable.html). However, this typically does not include versions of the dependencies.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- How exactly this should be enabled or disabled for a given target? Should there be a flag in the target configuration file, or some other mechanism?
+- How exactly should opt-in or opt-out from embedding this data be toggled? Should it be per-profile, like `strip = true`, or a global configuration option?
```

Comment:
Per profile, definitely. That will allow people to opt into shipping this information in specific build configurations, while omitting it in others.

Reply:
Does per-platform configuration make sense? Could we have it configured for *both* certain platforms and certain build profiles?

Reply:
Perhaps, but not more so than other profile settings. In general we would probably benefit from a mechanism to have target-specific profile settings, but I probably don't think we need that as a prerequisite for this.

Reply:
Thanks! I've updated the RFC to specify this.